### PR TITLE
Add function to profile list of transformers

### LINKED ISSUE #355: Add function to profile list of transformers
### Problem Description

The `performance` module should have a function that takes in a list of transformers and returns a `DataFrame` with the following columns:
- Transformer name (full name)
- Dataset generator name
- Num rows
- Fit time
- Transform time
- Reverse transform time
- Fit memory
- Transform memory
- Reverse Transform memory
- Commit #

By default, every transformer should be profiled.

### PR DESCRIPTION
**Pull request by amontanez24**
Resolves #355 
This PR:
- Adds a function called `evaluate_transformers` that takes a list of transformers and compiles a `DataFrame` summarizing their performance against different dataset generators. An example output of the `evaluate_transformers` function can be found below
<html>
<body>
<div class="cell code_cell rendered selected" tabindex="2" style="box-sizing: border-box; -webkit-box-orient: vertical; -webkit-box-align: stretch; display: flex; flex-direction: column; align-items: stretch; border-radius: 2px; border-width: 1px; border-style: solid; border-color: rgb(171, 171, 171); width: 1110px; padding: 5px; margin: 0px; outline: none; position: relative; overflow: visible; color: rgb(0, 0, 0); font-family: &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="output_wrapper" style="box-sizing: border-box; position: relative; -webkit-box-orient: vertical; -webkit-box-align: stretch; display: flex; flex-direction: column; align-items: stretch; z-index: 1;"><div class="output output_scroll" style="box-sizing: border-box; -webkit-box-orient: vertical; -webkit-box-align: stretch; display: block; flex-direction: column; align-items: stretch; height: 24em; width: 1098px; overflow: auto; border-radius: 2px; box-shadow: rgba(0, 0, 0, 0.8) 0px 2px 8px inset;"><div class="output_area" style="box-sizing: border-box; padding: 0px; break-inside: avoid; -webkit-box-orient: horizontal; -webkit-box-align: stretch; display: flex; flex-direction: row; align-items: stretch;"><div class="output_subarea output_html rendered_html output_result" dir="auto" style="box-sizing: border-box; color: rgb(0, 0, 0); overflow-x: visible; padding: 0.4em; -webkit-box-flex: 1; flex: 1 1 0%; max-width: calc(100% - 14ex);"><div style="box-sizing: border-box;">

Fit Time | Fit Memory | Transform Time | Transform Memory | Reverse Transform Time | Reverse Transform Memory | Number of fit rows | Number of transform rows | Dataset | Transformer
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
5.851000e-08 | 22.66800 | 1.089600e-07 | 41.54500 | 8.165500e-07 | 126.15900 | 1000.0 | 1000.0 | RandomBooleanGenerator | rdt.transformers.boolean.BooleanTransformer
5.148000e-09 | 3.16680 | 1.204300e-08 | 11.25450 | 1.090840e-07 | 37.18530 | 10000.0 | 10000.0 | RandomBooleanGenerator | rdt.transformers.boolean.BooleanTransformer
7.276000e-10 | 1.21668 | 1.641900e-09 | 8.32545 | 3.991120e-08 | 28.91853 | 100000.0 | 100000.0 | RandomBooleanGenerator | rdt.transformers.boolean.BooleanTransformer
8.861000e-08 | 27.30000 | 6.513500e-07 | 159.74300 | 2.413230e-06 | 181.99800 | 1000.0 | 1000.0 | RandomBooleanNaNsGenerator | rdt.transformers.boolean.BooleanTransformer
3.706000e-08 | 3.53000 | 1.482520e-07 | 52.78060 | 1.652570e-07 | 53.03410 | 10000.0 | 10000.0 | RandomBooleanNaNsGenerator | rdt.transformers.boolean.BooleanTransformer
... | ... | ... | ... | ... | ... | ... | ... | ... | ...
5.040775e-05 | 352.82470 | 2.428380e-05 | 338.80050 | 3.191700e-08 | 44.58350 | 10000.0 | 10000.0 | BigNormalGenerator | rdt.transformers.numerical.BayesGMMTransformer
6.121278e-05 | 330.48415 | 2.411937e-05 | 329.08005 | 2.075470e-08 | 44.05835 | 100000.0 | 100000.0 | BigNormalGenerator | rdt.transformers.numerical.BayesGMMTransformer
1.721808e-04 | 672.92900 | 2.600383e-05 | 496.58100 | 2.567600e-07 | 116.69100 | 1000.0 | 1000.0 | BigNormalNaNsGenerator | rdt.transformers.numerical.BayesGMMTransformer
4.995146e-05 | 376.49000 | 2.424078e-05 | 363.04150 | 4.244700e-08 | 47.57310 | 10000.0 | 10000.0 | BigNormalNaNsGenerator | rdt.transformers.numerical.BayesGMMTransformer
6.819039e-05 | 347.27357 | 2.434866e-05 | 345.90420 | 3.180970e-08 | 44.05835 | 100000.0 | 100000.0 | BigNormalNaNsGenerator | rdt.transformers.numerical.BayesGMMTransformer

<p style="box-sizing: border-box; margin: 1em 0px 0px; text-align: left;">346 rows × 10 columns</p></div></div></div></div></div></div><div class="cell code_cell rendered unselected" tabindex="2" style="box-sizing: border-box; -webkit-box-orient: vertical; -webkit-box-align: stretch; display: flex; flex-direction: column; align-items: stretch; border-radius: 2px; border-width: 1px; border-style: solid; border-color: transparent; width: 1110px; padding: 5px; margin: 0px; outline: none; position: relative; overflow: visible; color: rgb(0, 0, 0); font-family: &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="input" style="box-sizing: border-box; break-inside: avoid; -webkit-box-orient: horizontal; -webkit-box-align: stretch; display: flex; flex-direction: row; align-items: stretch;"><div class="prompt_container" style="box-sizing: border-box; display: flex; flex-direction: row; justify-content: space-between; align-items: flex-start; text-align: right;"><div class="prompt input_prompt" style="box-sizing: border-box; min-width: 15ex; padding: 0.4em; margin: 0px; font-family: monospace; text-align: right; line-height: 1.21429em; user-select: none; cursor: default; color: rgb(48, 63, 159); border-top: 1px solid transparent;"><bdi style="box-sizing: border-box;">In</bdi> [ ]:</div></div><div class="inner_cell" style="box-sizing: border-box; min-width: 0px; -webkit-box-orient: vertical; -webkit-box-align: stretch; display: flex; flex-direction: column; align-items: stretch; -webkit-box-flex: 1; flex: 1 1 0%;"><div class="input_area" aria-label="Edit code here" style="box-sizing: border-box; border: 1px solid rgb(207, 207, 207); border-radius: 2px; background: rgb(247, 247, 247); line-height: 1.21429em;"><div class="CodeMirror cm-s-ipython" style="box-sizing: border-box; font-family: monospace; height: auto; color: black; direction: ltr; position: relative; overflow: hidden; background: none; line-height: 1.21429em; font-size: 14px;"><div style="box-sizing: border-box; overflow: hidden; position: relative; width: 3px; height: 0px; top: 5.59375px; left: 4px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="box-sizing: border-box; color: inherit; font: inherit; margin: 0px; overflow: auto; position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scroll" tabindex="-1" style="box-sizing: content-box; margin-bottom: -50px; margin-right: -50px; padding-bottom: 50px; height: 11.1875px; outline: none; position: relative; overflow: auto hidden;"><div class="CodeMirror-sizer" style="box-sizing: content-box; position: relative; border-right: 33px solid transparent; margin-left: 0px; min-width: 7px; margin-bottom: -17px; min-height: 28px; padding-right: 0px; padding-bottom: 0px;"><div style="box-sizing: border-box; position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation" style="box-sizing: border-box; padding: 0.4em 0px; cursor: text; min-height: 1px;"><div role="presentation" style="box-sizing: border-box; position: relative; outline: none;"><div class="CodeMirror-measure" style="box-sizing: border-box; position: absolute; width: 993.062px; height: 0px; overflow: hidden; visibility: hidden;"></div><div class="CodeMirror-measure" style="box-sizing: border-box; position: absolute; width: 993.062px; height: 0px; overflow: hidden; visibility: hidden;"></div><div style="box-sizing: border-box; position: relative; z-index: 1;"></div><div class="CodeMirror-cursors" style="box-sizing: border-box; visibility: hidden; position: relative; z-index: 3;"><div class="CodeMirror-cursor" style="box-sizing: border-box; border-left: 1.4px solid black; border-right: none; width: 0px; position: absolute; pointer-events: none; left: 4px; top: 0px; height: 17px;"></div></div><div class="CodeMirror-code" role="presentation" style="box-sizing: border-box; outline: none;"><pre class=" CodeMirror-line " role="presentation" style="box-sizing: border-box; overflow: visible; font-family: inherit; font-size: inherit; display: block; padding: 0px 4px; margin: 0px; line-height: inherit; color: inherit; word-break: break-all; overflow-wrap: normal; background: transparent; border: 0px; border-radius: 0px; white-space: pre; z-index: 2; position: relative; -webkit-tap-highlight-color: transparent; font-variant-ligatures: contextual;"><span role="presentation" style="box-sizing: border-box; padding-right: 0.1px;"><span cm-text="" style="box-sizing: border-box;">​</span></span></pre></div></div></div></div></div></div></div></div></div></div></div>
</body>
</html>

### CODE REVIEW

**Comment by amontanez24** [comment]
@kveerama Can you look at the output columns and see if there is anything we should add or clarify?

**Review by katxiao — APPROVED** [review]
lgtm!

**Comment by kveerama** [comment]
> @kveerama Can you look at the output columns and see if there is anything we should add or clarify?

@amontanez24 looks good! One quick question: what are the units for time and memory ? Perhaps that can be added in column title?

**Comment by amontanez24** [comment]
> > @kveerama Can you look at the output columns and see if there is anything we should add or clarify?
> 
> @amontanez24 looks good! One quick question: what are the units for time and memory ? Perhaps that can be added in column title?

Ah, I believe it is in seconds and in bytes. I will add to the column name. This makes me realize a problem though. @katxiao in `contributing.py` we label the memory as `Mb / row`, but I believe the `_profile_memory` method returns in bytes. Is this a mistake that I should also fix? And if so, should we convert to Mb?

File: rdt/performance/performance.py [hunk 1]
```diff
@@ -57,6 +63,47 @@ def evaluate_transformer_performance(transformer, dataset_generator):
...
+
+    Args:
+        transformers (list<rdt.transformers.BaseTransformer>):
+            List of transformer classes to evaluate.
+
+    Returns:
+        pandas.DataFrame:
+            A table containing the times and memory usages for ``fit``, ``transform`` and
+            ``reverse_transform`` for each transformer provided against every dataset generator
+            that applies to it. The table also contains the latest commit number.
```
Comment (fealho) [inline rdt/performance/performance.py:93]: What do you mean by `The table contains the latest commit number.`?

**Review by fealho — APPROVED** [review]
Looks good 👍

**Comment by katxiao** [comment]
> > > @kveerama Can you look at the output columns and see if there is anything we should add or clarify?
> > 
> > 
> > @amontanez24 looks good! One quick question: what are the units for time and memory ? Perhaps that can be added in column title?
> 
> Ah, I believe it is in seconds and in bytes. I will add to the column name. This makes me realize a problem though. @katxiao in `contributing.py` we label the memory as `Mb / row`, but I believe the `_profile_memory` method returns in bytes. Is this a mistake that I should also fix? And if so, should we convert to Mb?
> 
> https://github.com/sdv-dev/RDT/blob/0db6a21f904e84d560f37cb5cd0c597dc46a5341/tests/contributing.py#L478

@amontanez24 That's interesting, I think it should be `b / row` then.

File: rdt/performance/performance.py (hunk 1 above)
Comment (amontanez24) [reply rdt/performance/performance.py:93]: good catch I need to remove this

**Pull request merged**